### PR TITLE
Fix #691 - Add Mongock support

### LINKED ISSUE #691: Spring Boot Database Migration: Mongock support

### PR DESCRIPTION
**Pull request by beatfreaker**
Fix #691 - Add Mongock support

### CODE REVIEW

**Comment by beatfreaker** [comment]
@pascalgrimaud can you please review this PR

**Comment by pascalgrimaud** [comment]
@beatfreaker : yes, will do it when possible. I'm a little bit late, as I have several PRs to review before this one. Just be patient :)

**Comment by pascalgrimaud** [comment]
@beatfreaker : can you resolve the conflict plz?

File: src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/domain/MongockDomainService.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      SOURCE,
+      "MongockDatabaseConfiguration.java",
+      getPath(MAIN_JAVA, packageNamePath, mongockConfigPath)
+    );
+  }
+
+  @Override
+  public void addChangelogJava(Project project) {
+    project.addDefaultConfig(PACKAGE_NAME);
+    String packageNamePath = project.getPackageNamePath().orElse(getPath("com/mycompany/myapp"));
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/domain/MongockDomainService.java:89]: use constants in DefaultConfig

File: src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/domain/MongockDomainService.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    String mongockDBMigrationPath = "technical/infrastructure/secondary/mongock/dbmigration";
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/domain/MongockDomainService.java:90]: use TECHNICAL_INFRASTRUCTURE_SECONDARY

File: src/main/resources/generator/server/springboot/database/mongock/InitialMigrationSetup.java.mustache [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      this.template = template;
+  }
+
+  @Execution
+  public void changeSet() {}
+
+  @RollbackExecution
+  public void rollback() {}
+
+}
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/database/mongock/InitialMigrationSetup.java.mustache:22]:
```suggestion
  public void rollback() {}
}
```

File: src/main/resources/generator/server/springboot/database/mongock/InitialMigrationSetup.java.mustache [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import io.mongock.api.annotations.RollbackExecution;
+import org.springframework.data.mongodb.core.MongoTemplate;
+
+@ChangeUnit(id="init", order = "001")
+public class InitialMigrationSetup {
+  private final MongoTemplate template;
+
+  public InitialMigrationSetup(MongoTemplate template) {
+      this.template = template;
+  }
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/database/mongock/InitialMigrationSetup.java.mustache:14]:
```suggestion
  public InitialMigrationSetup(MongoTemplate template) {
    this.template = template;
  }
```

File: src/test/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/application/MongockApplicationServiceIT.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Test
+  void shouldInit() {
+    // Given
+    ProjectDTO projectDTO = TestUtils.readFileToObject("json/chips.json", ProjectDTO.class);
+    if (projectDTO == null) {
+      throw new GeneratorException("Error when reading file");
+    }
+    projectDTO.folder(FileUtils.tmpDirForTest());
+    Project project = ProjectDTO.toProject(projectDTO);
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/application/MongockApplicationServiceIT.java:43]:
```suggestion
    ProjectDTO projectDTO = TestUtils.readFileToObject("json/chips.json", ProjectDTO.class).folder(tmpDirForTest());
    Project project = ProjectDTO.toProject(projectDTO);
```

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Comment by pascalgrimaud** [comment]
Very good job here, @beatfreaker 
I still have 3 other comments:
- plz add a package-info.java, defined as a BusinessContext
- plz add a new README.md (similar to what is done in maven/README.md)
- call the API in the generate.sh file, at the same level than mongodbapp

**Comment by beatfreaker** [comment]
Hello @pascalgrimaud 

currently the build is failing for `mongodbapp` execution, and I suspect it is because of following warning

```
Warning:  Rule violated for class tech.jhipster.fullapp.technical.infrastructure.secondary.mongock.dbmigration.InitialMigrationSetup: lines covered ratio is 0.00, but expected minimum is 1.00
Warning:  Rule violated for class tech.jhipster.fullapp.technical.infrastructure.secondary.mongock.MongockDatabaseConfiguration: lines covered ratio is 0.00, but expected minimum is 1.00
```

As these generated files does not have any logic so I don't know what should we write in unit tests, could you please help me with it ?

**Comment by pascalgrimaud** [comment]
@beatfreaker : sure, what do you prefer? Suggestion in comment or commit directly to your branch ?

**Comment by beatfreaker** [comment]
@pascalgrimaud I am okay with both,

If you have limited time then just add it here, I can put it in code and test it.

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Comment by beatfreaker** [comment]
> @beatfreaker : here the simple unit test you can use:
> 
> ```java
> @UnitTest
> @ExtendWith(MockitoExtension.class)
> class InitialMigrationSetupTest {
> 
>   @InjectMocks
>   InitialMigrationSetup initialMigrationSetup;
> 
>   @Test
>   void shouldChangeSet() {
>     assertThatCode(() -> initialMigrationSetup.changeSet()).doesNotThrowAnyException();
>   }
> 
>   @Test
>   void shouldRollback() {
>     assertThatCode(() -> initialMigrationSetup.rollback()).doesNotThrowAnyException();
>   }
> }
> ```

This helps, thank you

**Comment by pascalgrimaud** [comment]
Fine to remove it, in this initial migration file.
In the other migration files (for User for example), it will be there, so no worry about that

File: src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+This context is used to add [Mongock](https://mongock.io/) to an existing project:
+
+- Required dependency in `pom.xml`
+- Spring configuration class with `@EnableMongock`
+- Sample migration class marked as `@ChangeUnit`
+
+## Maintainers
+
+- [Pascal Grimaud](https://github.com/pascalgrimaud)
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/README.md:13]: Don't you want to take the lead on this part @beatfreaker ? :-)

Comment (beatfreaker) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/README.md:13]: I can, updating it :)

Comment (pascalgrimaud) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/dbmigration/mongock/README.md:13]: Yes plz put your name :-)

**Review by pascalgrimaud — APPROVED** [review]

**Comment by pascalgrimaud** [comment]
Thanks for your work here @beatfreaker :heart:

**Comment by beatfreaker** [comment]
@pascalgrimaud Bounty claimed at https://opencollective.com/generator-jhipster/expenses/74748

**Comment by pascalgrimaud** [comment]
@beatfreaker : approved. If possible, can you contribute to the front part too? Otherwise, the feature won't be available

**Comment by beatfreaker** [comment]
@pascalgrimaud yeah sure I can make respective changes on frontend too, will create new ticket to handle it.

**Pull request merged**